src/librustc_traits/dropck_outlives.rs

```diff
@@ -210,12 +210,18 @@ fn dtorck_constraint_for_ty<'tcx>(
             Ok::<_, NoSolution>(())
         })?,
 
-        ty::Closure(_, substs) => rustc_data_structures::stack::ensure_sufficient_stack(|| {
-            for ty in substs.as_closure().upvar_tys() {
-                dtorck_constraint_for_ty(tcx, span, for_ty, depth + 1, ty, constraints)?;
+        ty::Closure(_, substs) => {
+            if !substs.as_closure().is_valid() {
+                return Err(NoSolution);
```

Comment:
I think that this code can *also* just recursive on the `upvar_tys`, because the inference variable should be replaced by this point with a tuple